src/services/user/UsersServices.ts

```diff
@@ -61,11 +68,42 @@ export default class RegisterServices {
         const userDetailsRegistered = await this._modelDetails.create(userDetailsSerialized);
         this._logger('info', 'User details saved on database');
 
-        userRegistered._doc.inProgress = { status: 'wait_to_confirm', code: null };
-
         return {
             ...userRegistered._doc,
             details: userDetailsRegistered,
         };
     }
+
+    public async confirmCode(id: string, code: string): Promise<ConfirmCodeResponse> {
+        const userInfo = await this._model.findOne(id);
+
+        if (!userInfo)
+            throw new HttpRequestErrors({
+                message: 'User not found in database',
+                code: HttpStatusCode.NOT_FOUND,
+                name: getErrorName(HttpStatusCode.NOT_FOUND),
+            });
+
+        if (!userInfo.inProgress || userInfo.inProgress.code !== code)
+            throw new HttpRequestErrors({
+                message: 'Invalid code',
+                code: HttpStatusCode.BAD_REQUEST,
+                name: getErrorName(HttpStatusCode.BAD_REQUEST),
+            });
+
+        userInfo.inProgress.status = 'done';
+
+        const userUpdated = await this._model.update(id, userInfo);
+
+        if (!userUpdated?.inProgress)
```

Comment:
Se o id encontrou um usuário na linha 78, aqui ele também vai encontrar, é realmente necessário esse if?

Reply:
Isso é para verificar se retornou do update corretamente, typescript reclamou que o userUpdated poderia ser null. Coloquei como variável para ter certeza de que o status que retornar é o que de fato está no banco de dados.

Reply:
Colocar um as User no final da linha 96 não forçaria o TS a entender que não vai ser null? Esse if da linha 98 tá redundante com o if da linha 80.

Reply:
Colocar `as User` no final da linha 96 não resolve o problema do TS? Esse if ficou redundante.

Reply:
Não, porque a chave inProgress é opcional então ela pode vir como string ou undefined, as duas opções que vejo é manter o if ou remover e retornar 'done' direto ao invés do que é retornado do banco de dados.